File: week5/backtester.cpp
#include <functional>
#include <vector>
using namespace std;

double backtester(function<string(const vector<double>&, int, int)> strategy, const vector<double>& data, int window, double initial_cash = 1000) {
    double cash = initial_cash;
    bool holding = false;
    for (int i = 0; i < data.size(); i++) {
        string signal = strategy(data, i, window);
        if (signal == "BUY" && !holding) {
            cash -= data[i];
            holding = true;
        } else if (signal == "SELL" && holding) {
            cash += data[i];
            holding = false;
        }
    }
    return cash;
}

File: week4/backtester.py
def backtester(strategy, data, window, initial_cash=1000):
    cash = initial_cash
    holding = False

    for i in range(len(data)):
        signal = strategy(data, i, window)
        if signal == "BUY" and not holding:
            cash -= float(data[i])
            holding = True
        elif signal == "SELL" and holding:
            cash += float(data[i])
            holding = False
    
    return cash
    